### PR TITLE
S3faulttolerence - Put Object (Simple) API support - Part1

### PR DESCRIPTION
**Pull request by DPG17**
- Code changes in PUT api (scenario: simple new object; no object overwrite) to support S3 fault mode
- Disabled UT that are failing due to changes (which will be enabled or removed later while fixing ST/UT)

### CODE REVIEW

**Comment by nileshgovande** [comment]
Tagging @amitwac2608 @knrajnambiar76 for review

File: server/s3_motr_writer.h [hunk 1]
```diff
@@ -169,6 +172,17 @@ class S3MotrWiter {
     layout_ids.push_back(id);
   }
 
+  void set_buffer_rewrite_flag(bool is_buffer_re_write) {
```
Comment (amitwac2608) [inline server/s3_motr_writer.h:175]: make it inline function  !!!

File: server/s3_object_metadata.h [hunk 2]
```diff
@@ -257,7 +266,14 @@ class S3ObjectMetadata : private S3ObjectMetadataCopyable {
   std::string get_acl_as_xml();
   unsigned int get_number_of_parts() const { return this->obj_parts; }
   unsigned int get_number_of_fragments() const { return this->obj_fragments; }
+  void set_number_of_parts(unsigned int parts) { this->obj_parts = parts; }
```
Comment (amitwac2608) [inline server/s3_object_metadata.h:269]: can we make this and others as inline functions !!

File: server/s3_object_metadata.cc [hunk 3]
```diff
@@ -882,9 +928,14 @@ void S3ObjectExtendedMetadata::load(std::function<void(void)> on_success,
 void S3ObjectExtendedMetadata::get_obj_ext_entries(std::string last_object) {
   s3_log(S3_LOG_DEBUG, request_id, "Searching index from start key = [%s]\n",
          last_object.c_str());
+  unsigned int fetch_coumt = fragments;
```
Comment (amitwac2608) [inline server/s3_object_metadata.cc:931]: fetch_count

File: server/s3_put_object_action.cc [hunk 4]
```diff
@@ -58,7 +58,20 @@ S3PutObjectAction::S3PutObjectAction(
   old_object_oid = {0ULL, 0ULL};
   old_layout_id = -1;
   new_object_oid = {0ULL, 0ULL};
-
+  fault_mode_active = false;
```
Comment (amitwac2608) [inline server/s3_put_object_action.cc:61]: why not initialize them in the initializer list !!

File: server/s3_put_object_action.cc [hunk 5]
```diff
@@ -58,7 +58,20 @@ S3PutObjectAction::S3PutObjectAction(
...
+  last_object_size = 0;
+  primary_object_size = 0;
+  total_object_size_consumed = 0;
+  no_of_blocks_written = 0;
+  create_fragment_when_write_success = false;
+
+  // Default to 10 objects in S3 fault mode
+  max_objects_in_s3_fault_mode = 10;
```
Comment (amitwac2608) [inline server/s3_put_object_action.cc:69]: I think we should make this 10 value as const in the header file

File: server/s3_put_object_action.cc [hunk 6]
```diff
@@ -58,7 +58,20 @@ S3PutObjectAction::S3PutObjectAction(
...
+  current_fault_iteration = 0;
+  if (S3Option::get_instance()->get_max_objects_in_fault_mode() > 0) {
```
Comment (amitwac2608) [inline server/s3_put_object_action.cc:71]: should we put WARN or INFO message that the value specified in the config file is not within "unsigned short" range in else part ???

File: st/clitests/aws_iam_credential_file [hunk 7]
```diff
@@ -1,5 +1,3 @@
 [default]
-aws_access_key_id = G6OFkXkYTyS_YRmxenLbBg
-aws_secret_access_key = KPykpvZqBQFysZ8vkmGwrPLXyafnOJ52n/qpRVOF
-
-
+aws_access_key_id = AKIAZhILo7aqSKWuT_YLVtutaQ
```
Comment (amitwac2608) [inline st/clitests/aws_iam_credential_file:2]: this change should not be part of this PR

File: st/clitests/s3iamcli_test_config.yaml [hunk 8]
```diff
@@ -21,4 +21,4 @@
 
 # LDAP credentials used for creating s3 credentials
 ldapuser: 'sgiamadmin'
-ldappasswd: 'ldapadmin'
+ldappasswd: 'Admin@123'
```
Comment (amitwac2608) [inline st/clitests/s3iamcli_test_config.yaml:24]: I guess we do not want to have this file change as part of the PR !!!

File: server/s3_put_object_action.cc (hunk 6 above)
Comment (amitwac2608) [reply server/s3_put_object_action.cc:71]: Also can CORTX handle unsigned short highest value : 65k ??

If yes, we should ask QA to specifically test with the highest value.

otherwise can we put a check, like if value (in config file) is greater than N, then set the "max_objects_in_s3_fault_mode" to N

File: s3config.yaml.sample [hunk 9]
```diff
@@ -70,6 +70,7 @@ S3_SERVER_CONFIG:
    S3_SERVER_MOTR_ETIMEDOUT_MAX_THRESHOLD: 100          # Number of ETIMEDOUT errors per monitoring window before s3server restart
    S3_SERVER_MOTR_ETIMEDOUT_WINDOW_SEC: 1               # Monitoring window for motr ETIMEDOUT errors in seconds
    S3_SERVER_ENABLE_ADDB_DUMP: true                     # If set to true, then addb dump will be collected
+   S3_MAX_EXTENDED_OBJECTS_IN_FAULT_MODE: 10            # Maximum number of extended objects created when S3 write IO fails
```
Comment (knrajnambiar76) [inline s3config.yaml.sample:73]: Max number of allowed recovery by s3 when fault happens and motr error's out with specific error code

By default this will be 1
If its 0 then we dont do any recovery

File: server/s3_object_metadata.cc (hunk 3 above)
Comment (DPG17) [reply server/s3_object_metadata.cc:931]: ok

File: server/s3_put_object_action.cc (hunk 4 above)
Comment (DPG17) [reply server/s3_put_object_action.cc:61]: yes, will do in initialize list.

File: server/s3_put_object_action.cc (hunk 5 above)
Comment (DPG17) [reply server/s3_put_object_action.cc:69]: ok

File: server/s3_put_object_action.cc (hunk 6 above)
Comment (DPG17) [reply server/s3_put_object_action.cc:71]: We would certainly need conditioning in code, and if found greater than N, set it to N. 
Also, practically, we may end up supporting very less number of objects in fault mode, e.g., 3 or 5 or 10 or 30...etc. Moreover, @knrajnambiar76 recommends it to be 1, but I'm not sure.

File: st/clitests/aws_iam_credential_file (hunk 7 above)
Comment (DPG17) [reply st/clitests/aws_iam_credential_file:2]: oh...yes...

File: s3config.yaml.sample (hunk 9 above)
Comment (DPG17) [reply s3config.yaml.sample]: yes. If it is 0, the feature is disabled. This is how it is implemented. 
I'll add the text you mentioned "Max number of allowed recovery by s3 when fault happens and motr error's out with specific error code" in s3config file

File: st/clitests/s3iamcli_test_config.yaml (hunk 8 above)
Comment (DPG17) [reply st/clitests/s3iamcli_test_config.yaml]: Removed these changes, not required for PR.

File: server/s3_motr_writer.h [hunk 10]
```diff
@@ -102,6 +102,9 @@ class S3MotrWiter {
   // md5 for the content written to motr.
   MD5hash md5crypt;
 
+  // Flag to indicate re-write of same buffer, possibly to another object
+  bool re_write_buffer;
```
[the remark this answers is not in the document]
Comment (DPG17) [reply server/s3_motr_writer.h:106]: I think, "is_re_write_buffer" will be more readable.

[the remark this answers is not in the document]
Comment (DPG17) [reply server/s3_motr_writer.h:106]: I'll keep as it is now. If another reviewer has similar views, will change it.

File: server/s3_object_metadata.h (hunk 2 above)
Comment (DPG17) [reply server/s3_object_metadata.h:269]: ok. will make them inline

File: server/s3_object_metadata.cc (hunk 3 above)
Comment (DPG17) [reply server/s3_object_metadata.cc:931]: done.

File: server/s3_motr_writer.h (hunk 1 above)
Comment (DPG17) [reply server/s3_motr_writer.h:175]: done

File: s3config-test.yaml [hunk 11]
```diff
@@ -70,6 +70,7 @@ S3_SERVER_CONFIG:                                     # Section for S3 Server
    S3_SERVER_MOTR_ETIMEDOUT_MAX_THRESHOLD: 100          # Number of ETIMEDOUT errors per monitoring window before s3server restart
    S3_SERVER_MOTR_ETIMEDOUT_WINDOW_SEC: 2               # Monitoring window for motr ETIMEDOUT errors in seconds
    S3_SERVER_ENABLE_ADDB_DUMP: true                     # If set to true, then addb dump will be collected
+   S3_MAX_EXTENDED_OBJECTS_IN_FAULT_MODE: 0             # Maximum number of allowed recovery by S3 when fault happens and Motr error out with specific error code. When 0, feature is disabled.
```
Comment (knrajnambiar76) [inline s3config-test.yaml:73]: default will be 1

File: server/s3_common.h [hunk 12]
```diff
@@ -57,6 +57,10 @@
 #define MINIMUM_PART_NUMBER 1
 // Maxmimum part number for multipart operations
 #define MAXIMUM_PART_NUMBER 10000
+// Maximum number of allowed recovery by S3 when fault happens and Motr error
+// out with specific error code.
+// When 0, feature is disabled.
+#define MAX_ALLOWED_RECOVERY_IN_FAULT_MODE 10
```
Comment (knrajnambiar76) [inline server/s3_common.h:63]: In s3config yaml file this max value has to be mentioned

**Review by knrajnambiar76 — APPROVED** [review]

File: server/s3_common.h (hunk 12 above)
Comment (DPG17) [reply server/s3_common.h:63]: ok

File: s3config-test.yaml (hunk 11 above)
Comment (DPG17) [reply s3config-test.yaml:73]: default will be 0 (i.e. feature is disabled)

**Review by amitwac2608 — APPROVED** [review]

File: s3config.yaml.sample [hunk 13]
```diff
@@ -70,6 +70,7 @@ S3_SERVER_CONFIG:
    S3_SERVER_MOTR_ETIMEDOUT_MAX_THRESHOLD: 100          # Number of ETIMEDOUT errors per monitoring window before s3server restart
    S3_SERVER_MOTR_ETIMEDOUT_WINDOW_SEC: 1               # Monitoring window for motr ETIMEDOUT errors in seconds
    S3_SERVER_ENABLE_ADDB_DUMP: true                     # If set to true, then addb dump will be collected
+   S3_MAX_EXTENDED_OBJECTS_IN_FAULT_MODE: 0             # Maximum number of extended objects created when S3 write IO fails
```
[the remark this answers is not in the document]
Comment (DPG17) [reply s3config.yaml.sample]: no, (to the best of my knowledge)

[the remark this answers is not in the document]
Comment (amitwac2608) [reply s3config.yaml.sample]: Ok, we will discuss this later on this, for now we are good to merge this.
